### PR TITLE
[GPU/OpenCL] Initial version of Reshape Layer with OpenCL ops

### PR DESCRIPTION
**Pull request by niket-agarwal**
Added initial version of Reshape Layer for GPU. This is a basic implementation using naive kernel.

**Changes added with this PR:**

- `reshape_cl.cpp` added containing the new `ReshapeLayerCL` class for OpenCL implementation.
- Added `unittest_layers_reshape_cl.cpp` to test Reshape Layer on GPU.

Signed-off-by: Niket Agarwal [niket.a@samsung.com](mailto:niket.a@samsung.com)

### CODE REVIEW

**Comment by djeong20** [comment]
just wondering. why do we need this ReshapeLayer in the first place?
I think operations (e.g., forwarding, backwarding) would be a simple data copy that doesn't change any order of the data.

File: nntrainer/layers/cl_layers/reshape_cl.cpp [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        int input_index = batch_index * channels * height * width +
+                          input_channel * height * width +
+                          input_height * width +
+                          input_width;
+
+        int output_index = batch_index * elements_per_batch + element_index;
+        
+        output[output_index] = input[input_index];
+    }
+})";
```
Comment (djeong20) [inline nntrainer/layers/cl_layers/reshape_cl.cpp:78]: if the whole process is simply copying, couldn't it be as follows?

```
int i= get_global_id(0);
output[i] = input[i]
```

File: nntrainer/layers/cl_layers/reshape_cl.h [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  inline static const std::string type = "reshape";
+
+  static opencl::Kernel kernel_reshape;
+  static opencl::Kernel kernel_reshape_fp16;
+
+  /**
+   * @brief Process data and dimensions for reshape operation
+   * @param[in] input Tensor
+   * @param[in] result Tensor
+   * @param[in] RunLayerContext reference
```
Comment (SeoHyungjun) [inline nntrainer/layers/cl_layers/reshape_cl.h:110]: In the functions below, the same argument is written with a different explanation.

```suggestion
   * @param[in] context RunLayerContext reference
```

File: test/unittest/layers/unittest_layers_reshape_cl.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Niket Agarwal <niket.a@samsung.com>
+ *
+ * @file unittest_layers_reshape_cl.cpp
+ * @date 18th June 2024
+ * @brief Reshape Layer Test
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Niket Agarwal <niket.a@samsung.com>
+ * @bug No known bugs except for NYI items
```
Comment (SeoHyungjun) [inline test/unittest/layers/unittest_layers_reshape_cl.cpp:10]: The indentation of the code is slightly different.
In addition, it would be better to change the date format and the order of @brief for unity with other parts.

```suggestion
 * @file   unittest_layers_reshape_cl.cpp
 * @date   18 June 2024
 * @see    https://github.com/nnstreamer/nntrainer
 * @author Niket Agarwal <niket.a@samsung.com>
 * @bug    No known bugs except for NYI items
 * @brief  Reshape Layer Test
```

**Review by SeoHyungjun — APPROVED** [review]
LGTM

File: nntrainer/layers/cl_layers/reshape_cl.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
```
Comment (EunjuYang) [inline nntrainer/layers/cl_layers/reshape_cl.cpp:79]: This kernel only holds for the case, where `{dimsize,0,0}`.
If you intended this global-level parallelization only, @djeong20 's suggestion seems to be better.
If you find more generalized one, however, please update this kernel to use the indices you computed in this code block.

File: nntrainer/layers/cl_layers/reshape_cl.cpp (hunk 1 above)
Comment (niket-agarwal) [reply nntrainer/layers/cl_layers/reshape_cl.cpp]: Thanks for the suggestion! I have tested and updated in the latest commit.

File: test/unittest/layers/unittest_layers_reshape_cl.cpp (hunk 3 above)
Comment (niket-agarwal) [reply test/unittest/layers/unittest_layers_reshape_cl.cpp]: Updated.

File: nntrainer/layers/cl_layers/reshape_cl.cpp (hunk 4 above)
Comment (niket-agarwal) [reply nntrainer/layers/cl_layers/reshape_cl.cpp]: Updated with @djeong20's suggestion. Thanks.

**Review by djeong20 — COMMENTED** [review]
Appreciate the hard work! One last suggestion is renaming the reshape_cl kernel as copy_cl. Since the kernel is basically copying data, I think copy_cl better reflects the function of the kernel (the final result would be `ReshapeLayerCl` uses the `copy_cl` kernel).

File: test/unittest/layers/unittest_layers_reshape_cl.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+GTEST_PARAMETER_TEST(ReshapeGPU, LayerSemanticsGpu,
+                     ::testing::Values(semantic_reshape_gpu));
+
+auto reshape_basic_plain = LayerGoldenTestParamType(
+  nntrainer::createLayer<nntrainer::ReshapeLayerCl>, {"target_shape=-1"},
+  "2:3:3:3", "reshape.nnlayergolden",
+  LayerGoldenTestParamOptions::SKIP_CALC_DERIV |
+    LayerGoldenTestParamOptions::SKIP_CALC_GRAD |
+    LayerGoldenTestParamOptions::USE_INC_FORWARD,
+  "nchw", "fp32", "fp32");
```
Comment (djeong20) [inline test/unittest/layers/unittest_layers_reshape_cl.cpp:33]: Please add the fp16 test case and negative test case (maybe in a later PR?)

**Review by djeong20 — APPROVED** [review]
Great work! LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM
